flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/operation/Operation.java

```diff
@@ -0,0 +1,194 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.gateway.service.operation;
+
+import org.apache.flink.table.gateway.common.operation.OperationHandle;
+import org.apache.flink.table.gateway.common.operation.OperationStatus;
+import org.apache.flink.table.gateway.common.operation.OperationType;
+import org.apache.flink.table.gateway.common.results.OperationInfo;
+import org.apache.flink.table.gateway.common.results.ResultSet;
+import org.apache.flink.table.gateway.common.utils.SqlGatewayException;
+import org.apache.flink.table.gateway.service.result.ResultFetcher;
+import org.apache.flink.table.gateway.service.utils.SqlExecutionException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Supplier;
+
+/** Entity that contains the resource to run and the execution results. */
+public class Operation {
+
+    private static final Logger LOG = LoggerFactory.getLogger(Operation.class);
+
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+
+    private final OperationHandle operationHandle;
+
+    private final OperationType operationType;
+    private final boolean hasResults;
+    private OperationStatus status;
+
+    private final Callable<ResultFetcher> resultSupplier;
+
+    private Future<?> invocation;
+    private ResultFetcher resultFetcher;
+    private SqlExecutionException operationError;
+
+    public Operation(
+            OperationHandle operationHandle,
+            OperationType operationType,
+            Callable<ResultFetcher> resultSupplier) {
+        this.operationHandle = operationHandle;
+        this.status = OperationStatus.INITIALIZED;
+        this.operationType = operationType;
+        this.hasResults = true;
+        this.resultSupplier = resultSupplier;
+    }
+
+    void runBefore() {
+        updateState(OperationStatus.PENDING);
+    }
+
+    void runAfter() {
+        updateState(OperationStatus.FINISHED);
+    }
+
+    public void run(ExecutorService service) {
+        invocation =
+                service.submit(
+                        () -> {
+                            try {
+                                runBefore();
+                                updateState(OperationStatus.RUNNING);
+                                ResultFetcher fetcher = resultSupplier.call();
+                                writeLock(() -> resultFetcher = fetcher);
```

Comment:
I'm thinking that can we use `AtomicReference` for `status` which may avoid so many locks.